packages/form-handlers/src/shared.ts

```diff
@@ -1,34 +1,93 @@
-export type EmailContent = {
-  plainText: string;
+// Input data common for all handlers
+export type FormInfo = {
+  projectId: string;
+  projectDomain: string;
+  pageUrl: string;
+  formData: FormData;
+  email: string;
+  senderDomain?: string;
+};
+
+export type EmailInfo = {
+  from: string;
+  to: string;
+  txt: string;
+  /** Only the body content, requers wrapping into `<!DOCTYPE html><html><body>...</body></html>` */
   html: string;
   subject: string;
 };
 
-export const formDataToEmailContent = ({
+export type Result = { success: true } | { success: false; errors: string[] };
+
+export const getFormEntries = (formData: FormData) =>
+  [...formData.entries()].filter(([key]) => key.startsWith("ws--") === false);
+
+export const getFormId = (formData: FormData) => {
+  for (const [key, value] of formData.entries()) {
+    if (key === "ws--form-id") {
+      return value;
+    }
+  }
+};
+
+export const formToEmail = ({
   formData,
-  prefix = "There has been a new submission of your form:",
-  subject = "New form submission",
-}: {
-  formData: FormData;
-  prefix?: string;
-  subject?: string;
-}): EmailContent => {
-  const entries = [...formData.entries()];
-
-  const htmlRows = entries
-    .map(
-      ([key, value]) =>
-        `<tr><td><strong>${key}:</strong></td><td>${value}</td></tr>`
-    )
-    .join("");
-
-  const plainTextRows = entries
-    .map(([key, value]) => `${key}: ${value}`)
-    .join("\n");
+  pageUrl,
+  projectDomain,
+  email,
+  senderDomain = "webstudio.mail",
+}: FormInfo): EmailInfo => {
+  let html = `<p>There has been a new submission of your form at <a href="${pageUrl}">${pageUrl}</a>:</p>`;
+  let txt = `There has been a new submission of your form at ${pageUrl}:\n\n`;
+
+  html += "<table><tbody>";
+
+  for (const [key, value] of getFormEntries(formData)) {
+    html += `<tr><td><strong>${key}:</strong></td><td>${value}</td></tr>`;
+    txt += `${key}: ${value}\n`;
+  }
+
+  html += "</tbody></table>";
 
   return {
-    subject,
-    plainText: `${prefix}\n\n${plainTextRows}`,
-    html: `<!DOCTYPE html><html><body><p>${prefix}</p><table><tbody>${htmlRows}</tbody></table></body></html>`,
+    from: `${projectDomain}@${senderDomain}`,
+    to: email,
+    subject: `New form submission from ${projectDomain}`,
+    txt,
+    html,
   };
 };
+
+export const getResponseBody = async (
+  response: Response
+): Promise<{ text: string; json?: Record<string, unknown> }> => {
+  const text = await response.text();
+  try {
+    const json = JSON.parse(text);
+    return typeof json === "object" && json !== null
+      ? { json, text }
+      : { text };
+  } catch {
+    return { text: text === "" ? response.statusText : text };
+  }
+};
+
+export const getErrors = (
```

Comment:
The whole body funciton looks like zod object with transform, like

`z.union(z.object({error: z.string()}).blabla( v => [v.errors]), ...)` not sure this work

Reply:
Zod feels like overkill. I want to keep this package light